flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/MapToMapAndMultisetToMultisetCastRule.java

```diff
@@ -0,0 +1,200 @@
+                                != null;
+    }
+
+    /* Example generated code for MULTISET<INT> -> MULTISET<FLOAT>:
+    org.apache.flink.table.data.MapData _myInput = ((org.apache.flink.table.data.MapData)(_myInputObj));
+    boolean _myInputIsNull = _myInputObj == null;
+    boolean isNull$0;
+    org.apache.flink.table.data.MapData result$1;
+    float result$2;
+    isNull$0 = _myInputIsNull;
+    if (!isNull$0) {
+        java.util.Map map$838 = new java.util.HashMap();
+        for (int i$841 = 0; i$841 < _myInput.size(); i$841++) {
+            java.lang.Float key$839 = null;
+            java.lang.Integer value$840 = null;
+            if (!_myInput.keyArray().isNullAt(i$841)) {
+                result$2 = ((float)(_myInput.keyArray().getInt(i$841)));
+                key$839 = result$2;
+            }
+            value$840 = _myInput.valueArray().getInt(i$841);
+            map$838.put(key$839, value$840);
+        }
+        result$1 = new org.apache.flink.table.data.GenericMapData(map$838);
+        isNull$0 = result$1 == null;
+    } else {
+        result$1 = null;
+    }
+    return result$1;
+
+     */
+    @Override
+    protected String generateCodeBlockInternal(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            String returnVariable,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        final LogicalType innerInputKeyType;
+        final LogicalType innerInputValueType;
+
+        final LogicalType innerTargetKeyType;
+        final LogicalType innerTargetValueType;
+        if (inputLogicalType.is(LogicalTypeRoot.MULTISET)) {
+            innerInputKeyType = ((MultisetType) inputLogicalType).getElementType();
+            innerInputValueType = new IntType(false);
+            innerTargetKeyType = ((MultisetType) targetLogicalType).getElementType();
+            innerTargetValueType = new IntType(false);
+        } else {
+            innerInputKeyType = ((MapType) inputLogicalType).getKeyType();
+            innerInputValueType = ((MapType) inputLogicalType).getValueType();
+            innerTargetKeyType = ((MapType) targetLogicalType).getKeyType();
+            innerTargetValueType = ((MapType) targetLogicalType).getValueType();
+        }
+
+        final String innerTargetKeyTypeTerm = boxedTypeTermForType(innerTargetKeyType);
+        final String innerTargetValueTypeTerm = boxedTypeTermForType(innerTargetValueType);
+        final String keyArrayTerm = methodCall(inputTerm, "keyArray");
+        final String valueArrayTerm = methodCall(inputTerm, "valueArray");
+        final String size = methodCall(inputTerm, "size");
+        final String map = newName("map");
+        final String key = newName("key");
+        final String value = newName("value");
+
+        return new CastRuleUtils.CodeWriter()
+                .declStmt(className(Map.class), map, constructorCall(HashMap.class))
```

Comment:
I'm not sure if a `HashMap` is safe to use because not all internal data structures have proper `hashCode/equals`. By looking at `org.apache.flink.table.planner.codegen.calls.ScalarOperatorGens#generateMap`, I also don't see any `HashMap`. A safer solution sounds like reusing the array to array casting logic and apply it keyArray and valueArray? In general, we don't guarantee unique keys for maps.

Reply:
I think this is safe, as the input map is assumed to be valid and should not have multiple values for the same key anyway.

Reply:
A cast to `MAP<CHAR(1), X>` could produce multiple values. But after a offline discussion, I think it is safe to use a HashMap here and the problem is rather in the `ScalarOperatorGens#generateMap`.